File: macos-app/Lexis/Services/Drive/GoogleAuth.swift
import AppKit
import CryptoKit
import Foundation
import Security

/// Native OAuth 2.0 for a Google "Desktop app" client, using the PKCE +
/// loopback-redirect flow Google requires for that client type. The user
/// creates their own OAuth client in their own Google Cloud project (see
/// README) and pastes its Client ID/Secret into Settings; nothing is baked
/// into this binary.
@MainActor
@Observable
final class GoogleAuth {
    static let scope = "https://www.googleapis.com/auth/drive.file"

    private(set) var isSignedIn: Bool
    private var accessToken: String?
    private var accessTokenExpiry: Date?

    private let secrets: SecretsStore

    init(secrets: SecretsStore = KeychainStore()) {
        self.secrets = secrets
        self.isSignedIn = secrets.get(.driveRefreshToken) != nil
    }

    func signIn(clientID: String, clientSecret: String) async throws {
        let verifier = Self.randomURLSafeString(length: 64)
        let challenge = Self.codeChallenge(forVerifier: verifier)
        let state = Self.randomURLSafeString(length: 16)

        let server = LoopbackRedirectServer()
        let port = try await server.start()
        let redirectURI = "http://127.0.0.1:\(port)/"

        guard var components = URLComponents(string: "https://accounts.google.com/o/oauth2/v2/auth") else {
            throw LexisError.driveSync("Couldn't build the Google sign-in URL.")
        }
        components.queryItems = [
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: Self.scope),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "access_type", value: "offline"),
            URLQueryItem(name: "prompt", value: "consent"),
            URLQueryItem(name: "state", value: state),
        ]
        guard let authURL = components.url else {
            throw LexisError.driveSync("Couldn't build the Google sign-in URL.")
        }

        NSWorkspace.shared.open(authURL)
        let result = try await server.waitForRedirect()

        guard result.state == state else {
            throw LexisError.driveSync("Google's sign-in response didn't match this request. Please try again.")
        }
        guard let code = result.code else {
            throw LexisError.driveSync(result.errorDescription ?? "Google sign-in was cancelled.")
        }

        let tokens = try await Self.exchangeCode(
            code: code, clientID: clientID, clientSecret: clientSecret,
            verifier: verifier, redirectURI: redirectURI
        )
        guard let refreshToken = tokens.refreshToken else {
            throw LexisError.driveSync("Google didn't return a refresh token. Try disconnecting any prior Lexis access at myaccount.google.com/permissions and signing in again.")
        }
        try secrets.set(refreshToken, for: .driveRefreshToken)
        accessToken = tokens.accessToken
        accessTokenExpiry = Date().addingTimeInterval(tokens.expiresIn)
        isSignedIn = true
    }

    func signOut() {
        try? secrets.delete(.driveRefreshToken)
        accessToken = nil
        accessTokenExpiry = nil
        isSignedIn = false
    }

    /// Returns a valid access token, refreshing via the stored refresh token if needed.
    func validAccessToken(clientID: String, clientSecret: String) async throws -> String {
        if let accessToken, let accessTokenExpiry, accessTokenExpiry > Date().addingTimeInterval(60) {
            return accessToken
        }
        guard let refreshToken = secrets.get(.driveRefreshToken) else {
            throw LexisError.driveSync("Not connected to Google Drive. Connect in Settings first.")
        }
        let tokens = try await Self.refresh(refreshToken: refreshToken, clientID: clientID, clientSecret: clientSecret)
        accessToken = tokens.accessToken
        accessTokenExpiry = Date().addingTimeInterval(tokens.expiresIn)
        isSignedIn = true
        return tokens.accessToken
    }

    // MARK: - Token exchange

    private struct TokenResponse: Decodable {
        var accessToken: String
        var expiresIn: Double
        var refreshToken: String?

        enum CodingKeys: String, CodingKey {
            case accessToken = "access_token"
            case expiresIn = "expires_in"
            case refreshToken = "refresh_token"
        }
    }

    private static func exchangeCode(
        code: String, clientID: String, clientSecret: String, verifier: String, redirectURI: String
    ) async throws -> (accessToken: String, refreshToken: String?, expiresIn: Double) {
        let params = [
            "code": code,
            "client_id": clientID,
            "client_secret": clientSecret,
            "code_verifier": verifier,
            "grant_type": "authorization_code",
            "redirect_uri": redirectURI,
        ]
        let decoded: TokenResponse = try await postToken(params: params)
        return (decoded.accessToken, decoded.refreshToken, decoded.expiresIn)
    }

    private static func refresh(
        refreshToken: String, clientID: String, clientSecret: String
    ) async throws -> (accessToken: String, expiresIn: Double) {
        let params = [
            "refresh_token": refreshToken,
            "client_id": clientID,
            "client_secret": clientSecret,
            "grant_type": "refresh_token",
        ]
        let decoded: TokenResponse = try await postToken(params: params)
        return (decoded.accessToken, decoded.expiresIn)
    }

    private static func postToken(params: [String: String]) async throws -> TokenResponse {
        var request = URLRequest(url: URL(string: "https://oauth2.googleapis.com/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = formEncode(params)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            let bodyText = String(data: data, encoding: .utf8) ?? ""
            throw LexisError.httpFailure(service: "Google", status: status, body: bodyText)
        }
        do {
            return try JSONDecoder().decode(TokenResponse.self, from: data)
        } catch {
            throw LexisError.decodingFailure(service: "Google", underlying: error.localizedDescription)
        }
    }

    private static func formEncode(_ params: [String: String]) -> Data {
        let allowed = CharacterSet.urlQueryAllowed.subtracting(CharacterSet(charactersIn: "+&="))
        let pairs = params.map { key, value -> String in
            let encodedKey = key.addingPercentEncoding(withAllowedCharacters: allowed) ?? key
            let encodedValue = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
            return "\(encodedKey)=\(encodedValue)"
        }
        return Data(pairs.joined(separator: "&").utf8)
    }

    // MARK: - PKCE

    private static func randomURLSafeString(length: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: length)
        _ = SecRandomCopyBytes(kSecRandomDefault, length, &bytes)
        return base64URLEncode(Data(bytes))
    }

    private static func codeChallenge(forVerifier verifier: String) -> String {
        let digest = SHA256.hash(data: Data(verifier.utf8))
        return base64URLEncode(Data(digest))
    }

    private static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: android-app/app/src/main/java/com/aboooooo57/lexume/network/GoogleAuth.kt
package com.aboooooo57.lexume.network

import android.content.Context
import android.content.Intent
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.aboooooo57.lexume.data.local.DriveOAuthConfig
import com.aboooooo57.lexume.data.local.SecretKey
import com.aboooooo57.lexume.data.local.SecureKeyStore
import com.aboooooo57.lexume.support.LexumeException
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.api.Scope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.Request
import org.json.JSONObject

/**
 * Native Google sign-in for Drive access - the Android analog of
 * `Services/Drive/GoogleAuth.swift`'s PKCE + loopback-redirect flow, but
 * genuinely simpler: Google Sign-In's own system UI (an Activity Google
 * Play Services provides) handles the whole consent screen, so there's no
 * local HTTP server to stand up (per the plan's own tech-stack notes).
 * Requesting `requestServerAuthCode` (with `forceCodeForRefreshToken`) still
 * yields a one-time code this class exchanges for a refresh token itself,
 * kept in [SecureKeyStore] - the same BYO-OAuth-client shape as macOS/iPad
 * (see [DriveOAuthConfig]), not a managed-backend token exchange.
 */
class GoogleAuth(private val secureKeyStore: SecureKeyStore) {
    var isSignedIn by mutableStateOf(false)
        private set

    private var accessToken: String? = null
    private var accessTokenExpiryMillis: Long = 0L

    /** Call once when a screen that cares about sign-in state appears - mirrors GoogleAuth.swift's `init` reading the Keychain synchronously, just as a suspend call instead. */
    suspend fun refreshSignedInState() {
        isSignedIn = secureKeyStore.get(SecretKey.DRIVE_REFRESH_TOKEN) != null
    }

    /** Builds the sign-in Intent for the caller to launch via `ActivityResultContracts.StartActivityForResult()`. */
    fun signInIntent(context: Context): Intent {
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestServerAuthCode(DriveOAuthConfig.WEB_CLIENT_ID, true)
            .requestScopes(Scope(DRIVE_SCOPE))
            .build()
        return GoogleSignIn.getClient(context, gso).signInIntent
    }

    /** Processes the result Intent from [signInIntent]'s launcher, exchanging the returned auth code for a refresh token. */
    suspend fun completeSignIn(data: Intent?): Unit = withContext(Dispatchers.IO) {
        val account = try {
            // getResult(ApiException::class.java), not the no-arg `.result`
            // property - only this typed overload throws ApiException
            // directly on failure; the no-arg one wraps it in a generic
            // RuntimeExecutionException instead.
            GoogleSignIn.getSignedInAccountFromIntent(data).getResult(ApiException::class.java)
        } catch (e: ApiException) {
            throw LexumeException.DriveSync(signInErrorMessage(e))
        }
        val authCode = account.serverAuthCode
            ?: throw LexumeException.DriveSync("Google didn't return an authorization code. Please try again.")
        val tokens = exchangeAuthCode(authCode)
        val refreshToken = tokens.refreshToken
            ?: throw LexumeException.DriveSync(
                "Google didn't return a refresh token. Try removing Lexume's access at " +
                    "myaccount.google.com/permissions and signing in again."
            )
        secureKeyStore.set(refreshToken, SecretKey.DRIVE_REFRESH_TOKEN)
        accessToken = tokens.accessToken
        accessTokenExpiryMillis = System.currentTimeMillis() + (tokens.expiresIn * 1000).toLong()
        isSignedIn = true
    }

    suspend fun signOut() {
        secureKeyStore.delete(SecretKey.DRIVE_REFRESH_TOKEN)
        accessToken = null
        accessTokenExpiryMillis = 0L
        isSignedIn = false
    }

    /** Returns a valid access token, refreshing via the stored refresh token if needed. */
    suspend fun validAccessToken(): String = withContext(Dispatchers.IO) {
        val cached = accessToken
        if (cached != null && accessTokenExpiryMillis > System.currentTimeMillis() + REFRESH_MARGIN_MILLIS) {
            return@withContext cached
        }
        val refreshToken = secureKeyStore.get(SecretKey.DRIVE_REFRESH_TOKEN)
            ?: throw LexumeException.DriveSync("Not connected to Google Drive. Connect in Settings first.")
        val tokens = refresh(refreshToken)
        accessToken = tokens.accessToken
        accessTokenExpiryMillis = System.currentTimeMillis() + (tokens.expiresIn * 1000).toLong()
        isSignedIn = true
        tokens.accessToken
    }

    // MARK: - Token exchange

    private data class TokenResponse(val accessToken: String, val expiresIn: Double, val refreshToken: String?)

    private suspend fun exchangeAuthCode(code: String): TokenResponse = postToken(
        mapOf(
            "code" to code,
            "client_id" to DriveOAuthConfig.WEB_CLIENT_ID,
            "client_secret" to DriveOAuthConfig.CLIENT_SECRET,
            "grant_type" to "authorization_code"
        )
    )

    private suspend fun refresh(refreshToken: String): TokenResponse = postToken(
        mapOf(
            "refresh_token" to refreshToken,
            "client_id" to DriveOAuthConfig.WEB_CLIENT_ID,
            "client_secret" to DriveOAuthConfig.CLIENT_SECRET,
            "grant_type" to "refresh_token"
        )
    )

    private suspend fun postToken(params: Map<String, String>): TokenResponse {
        val formBody = FormBody.Builder().apply {
            params.forEach { (key, value) -> add(key, value) }
        }.build()
        val request = Request.Builder()
            .url("https://oauth2.googleapis.com/token")
            .post(formBody)
            .build()
        val (statusCode, text) = HttpClients.shared.executeSuspending(request)
        if (statusCode !in 200..299) {
            throw LexumeException.HttpFailure("Google", statusCode, text.take(200))
        }
        val json = try {
            JSONObject(text)
        } catch (e: Exception) {
            throw LexumeException.DecodingFailure("Google", e.message ?: "malformed token response")
        }
        return TokenResponse(
            accessToken = json.getString("access_token"),
            expiresIn = json.getDouble("expires_in"),
            refreshToken = if (json.has("refresh_token")) json.getString("refresh_token") else null
        )
    }

    private fun signInErrorMessage(e: ApiException): String = if (e.statusCode == CommonStatusCodes.CANCELED) {
        "Google sign-in was cancelled."
    } else {
        "Google sign-in failed: ${e.message ?: e.statusCode}"
    }

    companion object {
        const val DRIVE_SCOPE = "https://www.googleapis.com/auth/drive.file"
        private const val REFRESH_MARGIN_MILLIS = 60_000L
    }
}
